### PR TITLE
Enhance help message generator

### PR DESCRIPTION
**Pull request by Zebradil**
First of all, thanks for the handy library. I'm using my fork of it for a long time already, but now I got some time and passion to prepare this PR. There were some more enhancements, but they're not polished enough and not obviously needed. 

This PR improves `help` target:
- support groups of targets (groups name starts from `^##`)
- sorts targets inside each group
- adds visual lines
- supports multiline target comments, usage:
  ```
  hello: ## Example hello world target
  hello: ## with multiline help message
  ```

Before:
```
Usage:
        make [flags...] [target...] [options...]

Flags:
        See make --help

Targets:
    hello                     example hello world target
    help                      Display this help
    self-update               update make-bootstrap

Options:
    VERBOSE=<mode>            Set mode to 1 for verbose output [default: 0]
```

After:
```
Usage:
    make [flags...] [target...] [options...]

Flags:
    See make --help

Targets:
  Test
    another ------------------ This should go first
    hello -------------------- Example hello world target
                               with multiline help message
  Built-in
    help --------------------- Display this help
    self-update -------------- Update make-bootstrap

Options:
    VERBOSE=<mode> Set mode to 1 for verbose output [default: 0]
```

### CODE REVIEW

File: .bootstrap.mk [hunk 1]
```diff
@@ -50,53 +50,75 @@ SHELL       := bash
...
 	printf "\n\033[0;33mOptions:\033[0m\n    \033[0;32m%-$${pad}s\033[0m Set mode to 1 for verbose output \033[0;33m[default: 0]\033[0m\n\n" 'VERBOSE=<mode>'
 .PHONY: help
 
 # Alternative `help` goals
 
 # Simple version using `sed`. Requires comment _before_ the target declaration
-## help: display this help
-# help: Makefile
-# 	@sed -n 's/^##//p' $<
+#     # help: display this help
```
Comment (dsiebel) [inline .bootstrap.mk]: It seems this breaks the sectioning so I would suggest we just remove this part completely or move it to the README as documentation.

File: .bootstrap.mk [hunk 2]
```diff
@@ -50,53 +50,75 @@ SHELL       := bash
...
-	done; \
-	\
-	readarray -t targets < <(for t in "$${targets[@]}"; do echo "$$t"; done | sort -d); \
-	\
-	for target in "$${targets[@]}"; do \
-		printf "    \033[0;32m%-$${pad}s\033[0m %s\n" "$${target%%:*}" "$${target##*## }"; \
-	done; \
-	\
+	readonly pad=$$(printf "%0.1s" "-"{1..25})
+	print_targets() {
```
Comment (dsiebel) [inline .bootstrap.mk]: Hey @Zebradil, thank you for your contribution, hope you are well :)

In a recent PR we introduced the trailing backslashes `\` on each line to keep make-bootstrap compatible with older make versions (see #2). In order to maintain compatibility would you mind re-adding them?

File: .bootstrap.mk [hunk 3]
```diff
@@ -50,53 +50,75 @@ SHELL       := bash
...
-		done < "$${mk}"; \
-	done; \
-	\
-	readarray -t targets < <(for t in "$${targets[@]}"; do echo "$$t"; done | sort -d); \
-	\
-	for target in "$${targets[@]}"; do \
-		printf "    \033[0;32m%-$${pad}s\033[0m %s\n" "$${target%%:*}" "$${target##*## }"; \
-	done; \
-	\
+	readonly pad=$$(printf "%0.1s" "-"{1..25})
```
Comment (dsiebel) [inline .bootstrap.mk]: Not sure yet whether I like the lines. Makes it less readable to my eyes. Let me think about it.

**Review by dsiebel — CHANGES_REQUESTED** [review]

File: .bootstrap.mk (hunk 2 above)
Comment (Zebradil) [reply .bootstrap.mk]: Sure, will do.

File: .bootstrap.mk (hunk 1 above)
Comment (Zebradil) [reply .bootstrap.mk]: Yeah, moving this to the README is a good idea.

File: .bootstrap.mk (hunk 3 above)
Comment (Zebradil) [reply .bootstrap.mk]: I see what you mean. I can think of two options right now:
- Change appearance of lines: use another symbol, less contrast color (gray) or both.
- Include an option to specify the help style, as far as there are several of the already.

Comment (dsiebel) [reply .bootstrap.mk]: Less contrast sounds like the best option.
Having the option to customize the styling is maybe a bit too much.

Comment (Zebradil) [reply .bootstrap.mk]: Here how it looks now in my environment:

image

File: .bootstrap.mk (hunk 2 above)
Comment (Zebradil) [reply .bootstrap.mk]: Done

File: .bootstrap.mk (hunk 1 above)
Comment (Zebradil) [reply .bootstrap.mk]: Moved and fixed.

**Comment by Zebradil** [comment]
@dsiebel why closed?

**Comment by Zebradil** [comment]
I see, the history is rewritten to change your email. I can create another PR if you're fine with the styling now (see the picture in the thread above).

**Comment by dsiebel** [comment]
> I see, the history is rewritten to change your email. I can create another PR if you're fine with the styling now (see the picture in the thread above).

Yeah, sorry about that. I didn't think about any open PRs :-/
Styling looks good to me 👍

**Pull request closed without merging**